### PR TITLE
[bugfix] Always return batch indices to prevent duplicated logic for the users

### LINKED ISSUE #9425: Gap in `IndexBatchSamplerWrapper` check for dataloaders used for prediction
## 🐛 Bug

Carrying forward the discussion here: https://github.com/PyTorchLightning/pytorch-lightning/issues/9380#issuecomment-915637292

The Trainer is currently inconsistent around checks for batch samplers. This code first checks if the dataloader has a batch sampler before applying the wrapper for prediction. https://github.com/PyTorchLightning/pytorch-lightning/blob/8407238d66df14c4476f880a6e6260b4bfa83b40/pytorch_lightning/trainer/data_loading.py#L161-L171

But this code in the prediction loop unconditionally accesses the attribute: https://github.com/PyTorchLightning/pytorch-lightning/blob/8407238d66df14c4476f880a6e6260b4bfa83b40/pytorch_lightning/loops/epoch/prediction_epoch_loop.py#L163-L164

This means if for any reason the `IndexBatchSamplerWrapper` wasn't applied, this access will fail.

Proposal for fix:
- Check if the dataloader has a batch sampler here before accessing the property: https://github.com/PyTorchLightning/pytorch-lightning/blob/8407238d66df14c4476f880a6e6260b4bfa83b40/pytorch_lightning/loops/epoch/prediction_epoch_loop.py#L163-L164

cc @ninginthecloud as this might occur when using iterables with the trainer

### Environment

- PyTorch Lightning Version (e.g., 1.3.0):
- PyTorch Version (e.g., 1.8)
- Python version:
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #9425

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by SeanNaren — APPROVED** [review]
LGTM!

File: CHANGELOG.md [hunk 1]
```diff
@@ -337,6 +337,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed freeing data iterators in loop `on_run_end` ([#9386](https://github.com/PyTorchLightning/pytorch-lightning/pull/9386))
 
 
+- Fixed `BasePredictionWriter` not returning the batch_indices in a non-distributed setting ([#9425](https://github.com/PyTorchLightning/pytorch-lightning/pull/9425))
```
Comment (awaelchli) [inline CHANGELOG.md:340]:
```suggestion
- Fixed `BasePredictionWriter` not returning the batch_indices in a non-distributed setting ([#9432](https://github.com/PyTorchLightning/pytorch-lightning/pull/9432))
```

File: pytorch_lightning/loops/epoch/prediction_epoch_loop.py [hunk 2]
```diff
@@ -160,8 +162,11 @@ def _build_kwargs(self, batch: Any, batch_idx: int, dataloader_idx: int) -> Dict
 
     def _store_batch_indices(self, dataloader_idx: int) -> None:
         """Stores the batch indices if the predictions should be stored."""
-        batch_sampler = self.trainer.predict_dataloaders[dataloader_idx].batch_sampler
+        current_dataloader = self.trainer.predict_dataloaders[dataloader_idx]
+        batch_sampler = getattr(current_dataloader, "batch_sampler", None)
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/prediction_epoch_loop.py:166]:
```suggestion
        batch_sampler = current_dataloader.batch_sampler
```
the batch sampler is an attribute of Dataloader

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/epoch/prediction_epoch_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py:166]: I thought so, but @ananthsub raised a concern about it.

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py:166]: @ananthsub what was your concern? usually he is on the side of stricter typing ^^

Comment (tchaton) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py:166]: And it might not be true if a user doesn't provide a DataLoader class.

Comment (edward-io) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py:166]: custom dataloaders (including our internal one) might not subclass from torch.utils.data.dataloader, but only implement the iterator interface.

Comment (ananthsub) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py:166]: @awaelchli , adding to @edward-io 's response, I think @tchaton 's fix here is the safest as it matches the current behavior around the batch sampler check here: https://github.com/PyTorchLightning/pytorch-lightning/blob/ec828b826717cd3b5beabcb6d0cacf41b2320a98/pytorch_lightning/trainer/data_loading.py#L160-L170

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py:166]: ok but note pretty much everywhere in lightning we are expecting an instance of DataLoader. I would be highly surprised if you can run Lightning without that. We also have no tests for non-dataloader data iterables

Comment (carmocca) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py:166]: +1 to @awaelchli concerns... Pretty much all of our code in `data_loading.py` assumes a `DataLoader` instance is passed.

Properly supporting non-dataloaders should be part of a different larger discussion. This  `getattr` call here just makes this piece a bit more confusing for those reading through the loops.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/loops/epoch/prediction_epoch_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py]: Done !
